pkg/kubelet/cm/cpumanager/policy_static.go

```diff
@@ -314,15 +315,200 @@ func (p *staticPolicy) updateCPUsToReuse(pod *v1.Pod, container *v1.Container, c
 	p.cpusToReuse[string(pod.UID)] = p.cpusToReuse[string(pod.UID)].Difference(cset)
 }
 
+// validatePodScopeResources checks for the "empty shared pool" scenario. This occurs
+// when the sum of exclusive container CPU requests consumes the entire pod-level
+// budget, leaving no CPUs for containers that require a shared pool. Such a
+// configuration is invalid because it would lead to containers in the shared
+// pool having an empty cpuset, causing them to run on the node's shared pool
+// and breaking NUMA affinity.
+func (p *staticPolicy) validatePodScopeResources(pod *v1.Pod) error {
+	// This will never happen, added to be explicit about this requirement
+	if p.affinity.GetTopologyScope() != topologymanager.PodTopologyScope {
+		return nil
+	}
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.PodLevelResourceManagers) || !resourcehelper.IsPodLevelResourcesSet(pod) {
+		return nil
+	}
+
+	podTotalCPUs, sumOfExclusiveContainerCPUs := p.getPodAndContainerCPUs(pod)
+
+	hasSharedContainers := false
+	for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
+		if p.guaranteedCPUs(pod, &container) == 0 {
+			hasSharedContainers = true
+			break
+		}
+	}
+
+	if hasSharedContainers && sumOfExclusiveContainerCPUs >= podTotalCPUs {
+		return fmt.Errorf("pod rejected, sum of exclusive container cpu requests equals pod budget, leaving no cpus for shared containers")
+	}
+
+	return nil
+}
+
+// getPodAndContainerCPUs calculates the total number of CPUs requested at the pod
+// level and the sum of CPUs requested by all guaranteed containers.
+func (p *staticPolicy) getPodAndContainerCPUs(pod *v1.Pod) (int, int) {
+	podTotalCPUs := p.podGuaranteedCPUs(pod)
+	sumOfExclusiveContainerCPUs := 0
+	for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
+		sumOfExclusiveContainerCPUs += p.guaranteedCPUs(pod, &container)
```

Comment:
likewise, this seems suspicious because we cobble together init container (which run sequentially) and app/sidecar container (which will run concurrently). Plus, it seems we can just extract `p.podGuaranteedCPUs()` and just move it in the calling sites.